File: src/main/java/com/github/joelws/release/tracker/endpoint/artifact/ArtifactEndpoint.java
package com.github.joelws.release.tracker.endpoint.artifact;

import com.github.joelws.release.tracker.interfaces.BusinessService;
import com.github.joelws.release.tracker.interfaces.ResourceEndPoint;

import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Response;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;

@Path("artifacts")
public class ArtifactEndpoint extends ResourceEndPoint<String> {
    public ArtifactEndpoint(BusinessService<String> service) {
        super(service);
    }

    @GET
    @Path("/")
    @Produces(APPLICATION_JSON)
    @Override
    public Response read(@QueryParam("artifact") String artifact) {
        return service.read(artifact);
    }

    @DELETE
    @Path("/")
    @Produces(APPLICATION_JSON)
    @Override
    public Response delete(@QueryParam("artifact") String artifact) {
        return service.delete(artifact);
    }
}

File: src/main/kotlin/com/joelws/release/tracker/endpoint/artifact/ArtifactEndpoint.kt
package com.joelws.release.tracker.endpoint.artifact

import com.joelws.release.tracker.interfaces.BusinessService
import com.joelws.release.tracker.interfaces.ResourceEndpoint
import ratpack.exec.Blocking
import ratpack.func.Action
import ratpack.handling.Chain

/*
Copyright 2016 Joel Whittaker-Smith

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/open class ArtifactEndpoint(service: BusinessService<String>) : ResourceEndpoint<String>(service) {
    override val endpoints: Action<Chain>
        get() = Action { chain ->
            chain
                    .path("artifact", { ctx ->
                        ctx
                                .byMethod { s ->
                                    s
                                            .get {
                                                val query = ctx.request.queryParams["query"]
                                                if (query == null) {
                                                    Blocking.get { service.list() }.then { generateResponse ->
                                                        generateResponse(ctx)
                                                    }
                                                } else {
                                                    service.read(query)(ctx)
                                                }
                                            }
                                            .post {
                                                ctx.request.body.then { body ->
                                                    Blocking.get { service.create(body.text) }.then { generateResponse ->
                                                        generateResponse(ctx)
                                                    }
                                                }
                                            }
                                            .delete {
                                                val query = ctx.request.queryParams["query"] ?: ""
                                                service.delete(query)(ctx)
                                            }
                                }

                    })

        }
}
